neuralmonkey/model/model_part.py

```diff
@@ -1,147 +1,114 @@
 """Basic functionality of all model parts."""
-
 from abc import ABCMeta
-from contextlib import contextmanager
-from typing import Any, Callable, Dict, List, Set, Tuple
+from typing import MutableSet, Set, List, Tuple
+
+from neuralmonkey.model.parameterized import Parameterized, InitializerSpecs
+from neuralmonkey.model.feedable import Feedable
+
+
+class GenericModelPart(metaclass=ABCMeta):
+    """Base class for Neural Monkey model parts.
+
+    Neural Monkey dynamically decides which model parts are in use when using a
+    specific trainer or a runner. Each trainer/runner holds a reference to a
+    top-level model part, which is then responsible for collecting references
+    to all `Parameterized` and `Feedable` objects that contribute to the
+    computation of its Tensors. This behavior is implemented using the
+    `get_dependencies` method, which is called recursively on all instances of
+    `GenericModelPart` class that are references from within a model part.
 
-import tensorflow as tf
+    Apart from the `get_dependencies` method, this class also provides two
+    properties: `list_dependencies` and `singleton_dependencies` which store
+    the names of the Python class attributes that are regarded as potential
+    dependents of the `GenericModelPart` object. These dependents are
+    automatically checked for type and when they are instances of the
+    `GenericModelPart` class, results of their `get_dependencies` are united
+    and returned as dependencies of the current object.
+    """
```

Comment:
Tenhleta konstrukce se mi moc nelíbí. IMHO by stačilo mít property, která se jmenuje dependencies, a když by ten atribut byl list nebo obecně iterable, tak by se proiteroval, tohle IHMO přidává cognitive load čtenářům těch model partů.

Kacíčská myšlenka: Dovedeme si představit situaci, kde je něco atribut třídy, ale z nějakého důvodu to není dependency. Pokud ne, tak bych prostě v `get_dependencies` iteroval přes všechny atributy třídy a zachoval bych se podle jejich typu.

Reply:
ad konstrukce: hotovo, je tam jen dependencies.
ad kacíři: Radši bych tam ty dependencies měl takhle napůl explicitně a třeba se na to detailně podíval v další refactor iteraci..